### DISCUSSION TITLE
Self-signed certificates for public API

### DISCUSSION THREAD

**Discussion by justinmoon** [fedimint/fedimint#1271] [General]
We currently use self-signed certs for connections between federation node.

But so far deployments have used letsencrypt to sign TLS certs. But this makes deployment harder for unsophisticated users. Could we use self-signed certs for public APIs that clients connect to?

If we made them part of the consensus config, and included a signature of the config by federation ID public key inside the "connect string" that user clients use to download configs from all guardians, then we think it would be secure.

Question: Will all our clients support this kind of custom certificate store?

**Comment by dpc** [comment]
Ping @Maan2003 @jkitman for awareness I guess.

**Reply by jkitman** [reply]
@justinmoon Once we have the endpoint to download signed client configs, we can include TLS certs or any other kind of connection info in there as well.

**Reply by maan2003** [reply]
I doubt it is possible on web/wasm. I am not sure how much we support wasm.

**Reply by dpc** [reply]
@justinmoon How would we verify? Who should we ask about it?

**Reply by justinmoon** [reply]
First step when someone joins federation is to use a public ["connection string" to download federation config](https://github.com/fedimint/fedimint/blob/e987bd52f7f8b49efa41f491ad19ca95b1c2fe95/client/cli/src/main.rs#L362-L375). The information in this "connection string" is basically trusted. Currently it's just the guardian public API endpoints, but in future it could have the threshold pubkey that @jkitman just added. If users trust contents of this connection string (e.g. pubkey) then they trust any configs signed by it.

In order to use self-signed certificates for API endpoints, you'd need to put the the TLS certs. 

The problem is that initial connection to the URLs in the "connection string". At that point they wouldn't trust any self-signed certs. Would it be OK to not use TLS for this first connection, but once we've downloaded a signed config then we start enforcing TLS? In order for a "man in the middle" to inject a malicious config they'd need to produce a valid signature with this key ... which only seems possible if they the re-used an old config.

**Reply by dpc** [reply]
> Would it be OK to not use TLS for this first connection, but once we've downloaded a signed config then we start enforcing TLS?

Both conditional TLS or extra port seem like kind of a meh solution.

Putting these TLS certs into connection string would probably baloon the connection string.


It might be possible to make a call with TLS and configure the client to ignore the certs. Then the client validates the config, and makes the same call again, this time with TLS certs checking from the previous config, to make sure everything still checks-out.

We could maybe fix the "old config attack", by putting a timestamp on the whole thing, and require guardians to re-sign once a day or something. Clients would reject it if it's older than that. Or 

A bit of extra complication, but might be worth it.

One place I worry about is if all the potential TLS clients allow a mode where they ignore cert verification completely. Most should, but some might not.
